### DISCUSSION TITLE
Client initial connection to server

### DISCUSSION THREAD

**Discussion by n-e-o-n-s** [miguelgrinberg/python-socketio#726] [Q&A]
I noticed that on the javascript version if we initialize our client first before our server, the client will simply wait for the server and connect. On the python version, if we initialize the client before the server it raises a "ConnectionError" exception. Seems like it only attempts to connect once.
`try:
            self.eio.connect(url, headers=headers, transports=transports,
                             engineio_path=socketio_path)
        except engineio.exceptions.ConnectionError as exc:
            self._trigger_event(
                'connect_error', '/',
                exc.args[1] if len(exc.args) > 1 else exc.args[0])
            raise exceptions.ConnectionError(exc.args[0]) from None`
 Is there already a feature where my client can wait for the server connection upon initial? I know the client reconnects after the server disconnected, but this only works if it was connected successfully on initial.  I implemented my own version of handling this, but i was just wondering if there was something available already and i just missed it.

**Comment by miguelgrinberg** [comment]
Yes, I did not build the connect feature to enter the reconnect loop if the initial connection fails, that only happens as a result of the connection dropping. So right now there is no workaround other than retrying yourself.

**Reply by mrodal** [reply]
why? it makes sense for the reconnection parameters to be used for the initial connection also

**Reply by miguelgrinberg** [reply]
Why? Because that seemed logical to me when I implemented this. The word "reconnection" implies there was a connection before, right?

In terms of implementation reconnection and retries on initial connection are handled in different way. I intended to implement reconnections and that's what I did. Retries can be implemented I guess, I just don't have a need, and nobody else came up with a good solution yet.

**Reply by mrodal** [reply]
Would you accept this solution?
![image](https://user-images.githubusercontent.com/20667544/158685358-e152f5c9-50c5-49db-a9a8-354dc4284b0a.png)

The `wait` parameter on connect looks like it could be used to enable that initial connection retrial if you prefer it as optional behavior, or we could add a new parameter.

Being that the js library does this by default I think it should also be default here to maintain the same behavior

**Reply by miguelgrinberg** [reply]
Is that all it takes? The reconnect function was designed to run as a background task, it never runs like this. I'm surprised it just works.

I'll have to think about the details. This probably needs a setting that defaults to disabled for backwards compatibility. And this is likely to break some tests thta'll have to be fixed, plus new test(s) to check that this works.

**Reply by mrodal** [reply]
Great! Let me know if I can be of help

**Comment by n-e-o-n-s** [comment]
> Yes, I did not build the connect feature to enter the reconnect loop if the initial connection fails, that only happens as a result of the connection dropping. So right now there is no workaround other than retrying yourself.

Cool, as of now i just have it on a while loop using the sleep function provided. Im sure using the event handlers is a better aproach but this will do, thank you very much
